community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -833,4 +802,123 @@ TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKerne
             return SCHEMA;
         }
     }
+
+    static class TransactionStatus
+    {
+        private static final int OPEN = 0;
+        private static final int CLOSING = 1;
+        private static final int CLOSED = 2;
+        private static final int SHUTDOWN = 3;
+        private static final int STATE_BITS_MASK = 0x3;
+        private static final int NON_STATE_BITS_MASK = 0xFFFF_FFFC;
+        private static final int TERMINATED = 1 << 3;
+
+        private static final AtomicIntegerFieldUpdater<TransactionStatus> statusUpdater =
+                AtomicIntegerFieldUpdater.newUpdater( TransactionStatus.class, "status" );
+        private volatile int status = CLOSED;
+        private Status terminationReason;
+
+        public void init()
+        {
+            statusUpdater.set( this, OPEN );
+            reset();
+        }
+
+        public void reset()
+        {
+            terminationReason = null;
+        }
+
+        public boolean isOpen()
+        {
+            return !isClosed();
+        }
+
+        public boolean isClosed()
+        {
+            return is( CLOSED );
+        }
+
+        public boolean isClosing()
+        {
+            return is( CLOSING );
+        }
+
+        public boolean isTerminated()
+        {
+            return (statusUpdater.get( this ) & TERMINATED) != 0;
+        }
+
+        public boolean terminate( Status reason )
+        {
+            int currentStatus;
+            do
+            {
+                currentStatus = statusUpdater.get( this );
+                if ( (currentStatus != OPEN) && (currentStatus != CLOSING) )
+                {
+                    return false;
+                }
+                terminationReason = reason;
+            }
+            while ( !statusUpdater.compareAndSet( this, currentStatus, currentStatus | TERMINATED ) );
+            return true;
+        }
+
+        public boolean closing()
```

Comment:
Looks like `closing()` and `shutdown()` have a lot in common, everything except some parameters, no